### DISCUSSION TITLE
Nuxt 3 - <Meta> and <Title> tags

### DISCUSSION THREAD

**Discussion by Hyperblaster** [nuxt/nuxt#17051] [Questions]
Hey people,

Nuxt 3 is awesome. Absolutely loving it, I have a question that i can't quite figure out.

If i have in my site:
```
useHead({
   title:'My site title',
   meta:[
    { 
      name:'og:title', 
      content:'My site title',
    }
   ],
})
```

On some pages i have the components like this
```
<Title>My Dynamic Page Title</Title>
<Meta name="og:title" content="My Dynamic Page Title"/>
```

What i'm hoping to do, is come up with a way that while those components are on the screen it uses those for the meta tags, and when they aren't it defaults back to what is in `useHead()`, The tricky part is that it adds both tags, and social networks like Facebook, Twitter, LinkedIn etc, use the first one in the DOM.

I've been wondering whether i could add some kind of `priority` or `weight` so that nuxt can understand which <Title> should be visible, and which <Meta> tags should be considered the priority.

I guess i could implement something like this in a service, but thought i'd get thoughts from the brains trust before i over-engineer this thing...

**Comment by huang-julien** [comment]
Hi, what about using events to modify a single useHead MetaObject ? 

Or combining it with a computed to return your MetaObject without deduplicated tags.

**Reply by Hyperblaster** [reply]
That's what i'm thinking, But that would mean i can't use the `<Meta>` tags provided by Nuxt right? As i don't think i can customise how they work.

But now i'm thinking i can probably write my own component that does the same thing right?

**Reply by huang-julien** [reply]
Yes you can do that. I think that you should have only one source  were you're controlling your meta tags. It would be a bit more complex to handle Head tags a multiple places in your app. 

I'm not very fond of using the <Head> tags of useHead/Nuxt, did you try to use provide/inject, so you will have only the page who's providing the MetaObject (or an object to loop over for your tags) and all child components of this page will be able to modify the object ? This is something i love to do.

**Reply by Hyperblaster** [reply]
Here is what i cobbled together as a solution, nothing super elegant or special but it's working for my needs atm.

**/plugins/meta.js**
```
// 
import { reactive, computed, ref } from 'vue';

export default defineNuxtPlugin(nuxtApp => {


    const meta = ref([]);
    const link = ref([]);
    const script = ref([]);
    const title = ref([]);

    ////////////////////////////////

    const source = reactive({
        title,
        meta,
        link,
        script,
    })

    const metaData = {
        source,
    }


    function clearUUID(refArray, entry) {
    	refArray.value = refArray.value.filter(function(e) {
    		return e.uuid !== entry.uuid;
    	});
    }
    ////////////////////////////////

    metaData.addTitle = function(entry) {
    	clearUUID(title, entry);
        title.value.push(entry);
    }

    metaData.addMeta = function(entry) {
    	clearUUID(meta, entry);
        meta.value.push(entry);
    }

    metaData.addLink = function(entry) {
    	clearUUID(link, entry);
        link.value.push(entry);
    }

    metaData.addScript = function(entry) {
    	clearUUID(script, entry);
        script.value.push(entry);
    }

    ////////////////////////////////

    metaData.removeTitle = function(uuid) {
        title.value = title.value.filter(function(entry) {
            return entry.uuid != uuid;
        });
    }

    metaData.removeMeta = function(uuid) {
        meta.value = meta.value.filter(function(entry) {
            return entry.uuid != uuid;
        });
    }

    metaData.removeLink = function(uuid) {
        link.value = link.value.filter(function(entry) {
            return entry.uuid != uuid;
        });
    }

    metaData.removeScript = function(uuid) {
        script.value = script.value.filter(function(entry) {
            return entry.uuid != uuid;
        });
    }

    function getLatestValue(array) {
    	const cleaned = getCleanedValues(array);
    	const candidate = cleaned[cleaned.length -1];

    	return candidate?.value;
    }

    function getCleanedValues(array, dedupeString) {
        const lookup = array.reduce(function(set, entry) {

            if (entry.disabled) {
                return set;
            }

            let key;

            if (dedupeString) {
                key = entry[dedupeString];
            } else {
                key = entry.uuid;
            }

            set[key] = entry;
            return set;
        }, {});

        return Object.values(lookup);
    }

    ////////////////////////////////

    metaData.output = computed(() => {
        const output = {};
        output.link = link.value;
        output.script = getCleanedValues(script.value, 'src');
        output.meta = getCleanedValues(meta.value, 'property');
        output.title = getLatestValue(title.value)

        return output;
    })

    ////////////////////////////////

    return {
        provide: {
            metaData,
        }
    }

})

```



Implemented like this
```

useHead($metaData.output);



 $metaData.addTitle({
        uuid:'route:title',
        value:`${routeTitle} - ${applicationTitle}`,
    });
    
    $metaData.addTitle({
        uuid:'article.title',
        value:`${dynamicArticle.title`,
    });
    
    
    
$metaData.addMeta({ uuid:'route:og:title', property:'og:title', content:routeTitle })
$metaData.addMeta({ uuid:'route:og:description', property:'og:description', content:routeDescription })

// Etc.

```


And then created extra components that can be added dynamically to a page:
```

// /components/HeadMeta.vue
<script setup>

import { defineProps, onMounted, onUnmounted, } from 'vue';
const { $metaData } = useNuxtApp();


const props = defineProps({
    uuid:String,
    charset: String,
    content: String,
    name: String,
    property: String,
    disabled:Boolean,
})

const data = {...props}
data.content = data.content.replace(/(<([^>]+)>)/gi, "");

    $metaData.addMeta(data)

onUnmounted(() => {
    $metaData.removeMeta(data)
})


</script>
```


The main idea being that each Meta Tag, Title, Link etc... should be provided with a `uuid` when it's added, ensuring that there will only ever be 1 instance of that tag based on `property` or `name` etc. Even if there are entries from multiple sources, and tags can be added/removed based on uuid. (building a multi tenanted Nuxt app, most of the code is generated and rendered at runtime rather than as routes built from files)
